sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -99,127 +107,297 @@ object Hash {
+      addString(variant).addTypedValue(v)
+
+    private def addTextMap(xs: SortedLookupList[Value]): this.type =
+      iterateOver(xs.toImmArray)((acc, x) => acc.addString(x._1).addTypedValue(x._2))
+
+    private def addList(xs: FrontStack[Value]): this.type =
+      iterateOver(xs.toImmArray)(_ addTypedValue _)
+
+    // we add non default filed together with there field 1-based index,
+    // we end using 0 delimiter.
+    def addRecord(value: ImmArray[(_, Value)]): this.type = {
+      var i = 1
+      def addField: this.type = addInt(i)
+      value.foreach { case (_, value) =>
+        value match {
+          case leaf: Value.ValueCidlessLeaf =>
+            leaf match {
+              case Value.ValueEnum(_, value) =>
+                // No default value for enum
+                discard(addField.addString(value))
+              case Value.ValueInt64(value) =>
+                if (value != 0)
+                  discard(addField.addLong(value))
+              case Value.ValueNumeric(value) =>
+                if (value.signum() != 0)
+                  discard(addField.addNumeric(value))
+              case Value.ValueText(value) =>
+                if (value.nonEmpty)
+                  discard(addField.addString(value))
+              case Value.ValueTimestamp(value) =>
+                if (value.micros != 0)
+                  discard(addField.addLong(value.micros))
+              case Value.ValueDate(value) =>
+                if (value.days != 0)
+                  discard(addField.addInt(value.days))
+              case Value.ValueParty(value) =>
+                // No default value for party
+                discard(addField.addString(value))
+              case Value.ValueBool(value) =>
+                if (value)
+                  discard(addField.addByte(1.toByte))
+              case Value.ValueUnit =>
+            }
+          case Value.ValueRecord(_, fields) =>
+            // No default value for records
+            discard(addField.addRecord(fields))
+          case Value.ValueVariant(_, variant, value) =>
+            // No default value for variant
+            discard(addField.addVariant(variant, value))
+          case Value.ValueContractId(value) =>
+            // No default value for contractId
+            discard(addField.addCid(value))
+          case Value.ValueList(values) =>
+            if (values.nonEmpty)
+              discard(addField.addList(values))
+          case Value.ValueOptional(value) =>
+            value match {
+              case Some(value) =>
+                discard(addField.addTypedValue(value))
+              case None =>
+            }
+          case Value.ValueTextMap(value) =>
+            if (!value.isEmpty)
+              discard(addTextMap(value))
+          case Value.ValueGenMap(entries) =>
+            if (entries.nonEmpty)
+              discard(addGenMap(entries))
+        }
+        i += 1
+      }
+      addInt(0)
```

Comment:
Since we never ship these encodings, but only hash it, I don't think the trouble with varints is worth the complexity. `0xFF` sounds more interesting though if you really want to save on a few bytes. In any case, I'd recommend to capture the desired properties in a high-level comment. (I'm currently trying to formulate the property that I think we want and get back to you when I have something that makes sense to me.)